### PR TITLE
Fix for pivot: error when 'values' is a multicharacter string

### PR DESCRIPTION
**Pull request by shaswat-indian**
## Description
Resolves: https://github.com/rapidsai/cudf/issues/10529

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/tests/test_reshape.py [hunk 1]
```diff
@@ -621,3 +621,32 @@ def test_crosstab_simple():
...
+        ["C", "b", 1, 1, 1],
+        ["C", "c", 2, 2, 2],
+    ]
+    columns = ["x", "y", "z", "z123", "123z"]
+    pdf = pd.DataFrame(data, columns=columns)
+    cdf = cudf.DataFrame(data, columns=columns)
+    expected = pd.pivot(pdf, index="x", columns="y", values=values)
+    actual = cudf.pivot(cdf, index="x", columns="y", values=values)
+    if not isinstance(values, (list, tuple)):
+        actual.columns = actual.columns.droplevel(0)
```
Comment (bdice) [inline python/cudf/cudf/tests/test_reshape.py:647]: Why is this needed? This looks like an incompatibility between pandas and cudf that we should resolve, but I don't know what this line does.

File: python/cudf/cudf/tests/test_reshape.py [hunk 2]
```diff
@@ -621,3 +621,32 @@ def test_crosstab_simple():
     expected = pd.crosstab(a, [b, c], rownames=["a"], colnames=["b", "c"])
     actual = cudf.crosstab(a, [b, c], rownames=["a"], colnames=["b", "c"])
     assert_eq(expected, actual, check_dtype=False)
+
+
+@pytest.mark.parametrize(
+    "values", ["z", "z123", ["z123"], ["z", "z123", "123z"]]
+)
+def test_pivot_values(values):
```
Comment (bdice) [inline python/cudf/cudf/tests/test_reshape.py:629]: I'd prefer to see this test placed next to the other `pivot` tests instead of at the bottom, below the `crosstab` tests.

File: python/cudf/cudf/tests/test_reshape.py (hunk 1 above)
Comment (shaswat-indian) [reply python/cudf/cudf/tests/test_reshape.py:647]: I noticed that presently even when `values` is not a list but just a single character string(which was supported even before this change) we were returning a multiindex dataframe unlike Pandas. Assuming this is the expected behavior as it already existed, I added this in the test.

For example, if `values='z'`
This returns a single index DataFrame in Pandas but a MultiIndex DataFrame in cuDF

But if `values=[z]`
This returns a MultiIndex DataFrame in both Pandas and cuDF.

Comment (bdice) [reply python/cudf/cudf/tests/test_reshape.py:647]: Yup! This is actually part of the same bug you're trying to solve, I think. Strings are iterables of characters, but we want to handle strings as standalone objects, to match pandas.

You might need something like:
```python
if not isinstance(values, (list, tuple, str)):
    values = [values]
```
to prevent `'z'` from being converted to `['z']` and becoming a MultiIndex.

Comment (shaswat-indian) [reply python/cudf/cudf/tests/test_reshape.py:647]: I've made a similar change for the multicharacter string issue but I don't think that affects the MultiIndex issue as the dataframe gets created in a different helper function called `_pivot` which is also used in `unstack`. I'll take a look if that affects the expected behavior in `unstack` as well.

File: python/cudf/cudf/tests/test_reshape.py (hunk 2 above)
Comment (shaswat-indian) [reply python/cudf/cudf/tests/test_reshape.py]: Done.

File: python/cudf/cudf/tests/test_reshape.py (hunk 1 above)
Comment (shaswat-indian) [reply python/cudf/cudf/tests/test_reshape.py]: I've made the change in `pivot` itself as unlike it for `unstack`, Pandas returns a MultiIndex dataframe itself even for string inputs.

File: python/cudf/cudf/core/reshape.py [hunk 3]
```diff
@@ -978,7 +982,10 @@ def pivot(data, index=None, columns=None, values=None):
     if len(columns_index) != len(columns_index.drop_duplicates()):
         raise ValueError("Duplicate index-column pairs found. Cannot reshape.")
 
-    return _pivot(values, index, columns)
+    result = _pivot(values, index, columns)
+    if not values_is_list:
+        result.columns = result.columns.droplevel(0)
```
Comment (shwina) [inline python/cudf/cudf/core/reshape.py:987]: If possible, we should do this by operating on the `ColumnAccessor` (`result._data`), rather  than using the `.columns` attribute (which is slow).

Comment (shaswat-indian) [reply python/cudf/cudf/core/reshape.py]: Done.

**Review by bdice — APPROVED** [review]
I think this looks like an appropriate fix. I don't know if there's a better way to handle the `MultiIndex` to `Index` code, but @shwina might know.

File: python/cudf/cudf/core/reshape.py [hunk 4]
```diff
@@ -983,8 +983,13 @@ def pivot(data, index=None, columns=None, values=None):
         raise ValueError("Duplicate index-column pairs found. Cannot reshape.")
 
     result = _pivot(values, index, columns)
+
+    # MultiIndex to Index
     if not values_is_list:
-        result.columns = result.columns.droplevel(0)
+        column_name = result._data.level_names[1]
+        table_columns = [multi_col[1] for multi_col in result._data.names]
+        result.columns = cudf.Index(data=table_columns, name=column_name)
```
Comment (shwina) [inline python/cudf/cudf/core/reshape.py:991]: We should really avoid touching the `.columns` attribute altogether. And we should _definitely_ avoid setting it  to a cuDF object as that will force a device-to-host copy. See the logic in the columns [setter](https://github.com/rapidsai/cudf/blob/127d57485445001e5d0c9fdc20fd8e380d9a410d/python/cudf/cudf/core/dataframe.py#L2244) method.

I think there's a bug in the `ColumnAccessor` droplevels function that makes it difficult to use here. Happy  to chat more about this offline, and we can see if including a fix  for that in this PR is appropriate.

Comment (shaswat-indian) [reply python/cudf/cudf/core/reshape.py]: I think I found the issue with `droplevel`. `names` is a cached property under `ColumnAccessor` but we expect it to change when a level is dropped. I've done a [fix](https://github.com/rapidsai/cudf/pull/11538/commits/788acff0404a293db685545a83c8b0002081cca2#diff-3541948bd042d45b52e62ae79b29555eae9dc2d00e060f058818c0ebb0c0b44dR619) for it and things seem to be working fine now. Let me know if something else needs to be changed.

**Comment by shaswat-indian** [comment]
rerun tests

**Review by shwina — APPROVED** [review]

**Comment by shwina** [comment]
@gpucibot merge
